tree/ntuple/v7/src/RNTupleSerialize.cxx

```diff
@@ -713,7 +713,7 @@ std::uint32_t ROOT::Experimental::Internal::RNTupleSerializer::SerializeFeatureF
          if (i == (flags.size() - 1))
             SerializeInt64(flags[i], bytes);
          else
-            bytes += SerializeInt64(-flags[i], bytes);
+            bytes += SerializeInt64(flags[i] | 0x8000000000000000, bytes);
```

Comment:
Don't you mean:
```suggestion
            bytes += SerializeInt64(flags[i] & ~0x8000000000000000, bytes);
```
(to set the most significant bit to zero?

And wouldn't:
```suggestion
     for (unsigned i = 0; i < flags.size() - 1; ++i) {
         if (flags[i] < 0)
            throw RException(R__FAIL("feature flag out of bounds"));
         bytes += SerializeInt64(flags[i] & ~0x8000000000000000,, bytes);
     }
     SerializeInt64(flags[i] & ~0x8000000000000000,, bytes);
```
be more straightforward (and independently I added a correction since it applies the transformation for all elements?)

Reply:
No, it is correct as-is: according to the specification, it should set the most significant bit to _one_ for all except the last 64-bit integer.

For the second suggestion: moving that `SerializeInt64()` call out of the loop would fail if the input vector `flags` has zero elements.

Reply:
> It is correct as-is:

I see.  You may want to still refer to that document (and/or write the gist of it) here (eg. I did not re-find that paragraph when trying to check the code).

> fail if the input vector flags has zero elements.

But it does not get there thanks to line 703/704

Reply:
> > It is correct as-is:
> 
> I see. You may want to still refer to that document (and/or write the gist of it) here (eg. I did not re-find that paragraph when trying to check the code).
> 
> > fail if the input vector flags has zero elements.
> 
> But it does not get there thanks to line 703/704

You are right, but in that case, we would not be checking if the last element in the vector has the MSb set by the caller (i.e., has a negative value).
For instance, the call `SerializeFeatureFlags(std::vector<std::int64_t>{0, -1}, buffer)` should throw (the `ntuple-serialize` unit test fails if it doesn't).

Adding another check for that after the loop is not much clearer.  Therefore, I would vote for leaving it as-is.

Reply:
> > It is correct as-is:
> 
> I see. You may want to still refer to that document (and/or write the gist of it) here (eg. I did not re-find that paragraph when trying to check the code).
> 
> > fail if the input vector flags has zero elements.
> 
> But it does not get there thanks to line 703/704

:+1: Done.

Reply:
```suggestion
    for (unsigned i = 0; i < flags.size() - 1; ++i) {
         if (flags[i] < 0)
            throw RException(R__FAIL("feature flag out of bounds"));
         bytes += SerializeInt64(flags[i] | 0x8000000000000000,, bytes);
    }
    if (flags[i] < 0)
        throw RException(R__FAIL("feature flag out of bounds"));
    SerializeInt64(flags[i], bytes);
    // or SerializeInt64(flags[i] & ~0x8000000000000000, bytes);
```
This version does indeed have a 2 lines duplication but is also somewhat faster (avoid the 'one more if per iteration) without being overly too complicated.

With the alternative (`// or ...`), the `if (flags[i] < 0)` are optional and could be move somewhere else (at the assignment of the flag?).  [Whether this is useful also depends on whether the `SerializeFeatureFlags` is an internal interface or is except to be called explicitly by users .. and of course the ratio of setting vs writing]

Reply:
For reference after merge: `SerializeFeatureFlags()` is an internal function called as part of header/footer serialization (2 times per written RNTuple).
Also given that the input vector is not expected to have more than 2-3 entries, this is not a hot-path and is therefore left as-is.